temporalcli/commands.workflow.go

```diff
@@ -309,3 +377,147 @@ func startBatchJob(cctx *CommandContext, cl client.Client, req *workflowservice.
 	cctx.Printer.Printlnf("Started batch for job ID: %v", req.JobId)
 	return nil
 }
+
+func getResetEventIDByType(ctx context.Context, resetType, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (string, int64, error) {
+	switch resetType {
+	case "LastWorkflowTask":
+		return getLastWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	case "LastContinuedAsNew":
+		return getLastContinueAsNewID(ctx, namespace, wid, rid, wfsvc)
+	case "FirstWorkflowTask":
+		return getFirstWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	default:
+		return "", -1, fmt.Errorf("invalid reset type: %s", resetType)
+	}
+}
+
+// Returns event id of the last completed task or id of the next event after scheduled task.
+func getLastWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace: namespace,
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+
+	for {
+		resp, err := wfsvc.GetWorkflowExecutionHistoryReverse(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				break
+			} else if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				// if there is no task completed event, set it to first scheduled event + 1
+				workflowTaskEventID = e.GetEventId() + 1
+			}
+		}
```

Comment:
You should extract a function which returns workflowTaskEventID

Reply:
See my comments on the other variants of this logic